modGui/src/org/aion/os/KernelLauncher.java

```diff
@@ -0,0 +1,226 @@
+                LOGGER.error("Found old kernel pid file at {}, but failed to deserialize it, " +
+                                "so it was ignored.", pidFile.getAbsolutePath(), ex);
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Kill the OS process that the kernel is running in.
+     *
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public void terminate() throws KernelControlException {
+        // Implemented by calling UNIX `kill` program on the pid of the process we launched.
+        // For the future, should add a terminate function in Aion API that we can call in
+        // order for this to be less platform-dependent and more reliable.
+        if(!hasLaunchedInstance()) {
+            throw new IllegalArgumentException("Trying to terminate when there is no running instance");
+        }
+
+        ProcessBuilder processBuilder = new ProcessBuilder()
+                .command("kill",
+                        String.valueOf(currentInstance.getPid()))
+                .redirectOutput(ProcessBuilder.Redirect.INHERIT)
+                .redirectError(ProcessBuilder.Redirect.INHERIT);
+        LOGGER.debug("About to kill pid {}", currentInstance.getPid());
+        try {
+            int killExitCode = processBuilder.start().waitFor();
+            LOGGER.trace("`kill` return code: " + killExitCode);
+            removePersistedPid();
+        } catch (InterruptedException | IOException ex) {
+            String message = "Error killing the Aion kernel process.";
+            LOGGER.error(message, ex);
+            throw new KernelControlException(message, ex);
+        }
+
+        setCurrentInstance(null);
+    }
+
+    /**
+     * Whether the launcher is associated with a running kernel instance.  This includes instances
+     * that KernelLauncher recovered through {@link #tryResume()}.
+     *
+     * @return whether the launcher is associated with a running kernel instance
+     */
+    public boolean hasLaunchedInstance() {
+        return currentInstance != null;
+    }
+
+    private KernelInstanceId retrieveAndSetPid(File pidFile) throws IOException, ClassNotFoundException {
+        FileInputStream fis = new FileInputStream(pidFile);
+        ObjectInputStream ois = new ObjectInputStream(fis);
+        return (KernelInstanceId) ois.readObject();
+    }
+
+    private KernelInstanceId setAndPersistPid(long pid) throws IOException {
+        KernelInstanceId kernel = new KernelInstanceId(pid);
+        FileOutputStream fos = new FileOutputStream(pidFile);
+        ObjectOutputStream oos = new ObjectOutputStream(fos);
+        oos.writeObject(kernel);
+        setCurrentInstance(kernel);
+        return kernel;
+    }
+
+    @VisibleForTesting void removePersistedPid() throws IOException {
+        if(pidFile.exists() && pidFile.isFile()) {
+            pidFile.delete();
```

Comment:
should catch the exception when deleting the file.

Reply:
Ok, will change